### PR TITLE
Ensures overloads are ordered from narrow to broad

### PR DESCRIPTION
**Pull request by Michael0x2a**
This commit reorders any overloads where the first overload was "shadowing" the second, preventing it from ever being matched by type checkers that work by selecting the first matching overload alternative.

For example, the first overload alternative below is strictly broader then the second, preventing it from ever being selected:

    class Parent: pass
    class Child(Parent): pass

    @overload
    def foo(x: *int) -> Parent: ...
    @overload
    def foo(x: int, y: int) -> Child: ...

The correct thing to do is to either delete the second overload or rearrange them to look like this:

    @overload
    def foo(x: int, y: int) -> Child: ...
    @overload
    def foo(x: *int) -> Parent: ...

Rationale: I'm currently [working on a proposal][0] that would amend PEP 484 to (a) mandate type checkers check overloads in order and (b) prohibit overloads where an earlier alternative completely shadows a later one.

  [0]: python/typing#253 (comment)

This would prohibit overloads that look like the example below, where the first alternative completely shadows the second.

I figured it would be a good idea to make these changes ahead of time: if my proposal is accepted, it'd make the transition smoother. If not, this is hopefully a relatively harmless change.

Note: I think some of these overloads could be simplified (e.g. `reversed(...)`), but I mostly stuck with rearranging them in case I was wrong. The only overload I actually changed was `hmac.compare_digest` -- I believe the Python 2 version actually accepts unicode.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
I assume you mean "overloads", not "commits" :)

File: stdlib/2/os/__init__.pyi [hunk 1]
```diff
@@ -219,9 +219,9 @@ def renames(old: _PathType, new: _PathType) -> None: ...
 def rmdir(path: _PathType) -> None: ...
 def stat(path: _PathType) -> Any: ...
 @overload
-def stat_float_times(newvalue: bool = ...) -> None: ...
-@overload
 def stat_float_times() -> bool: ...
+@overload
+def stat_float_times(newvalue: bool = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/os/__init__.pyi]: Shouldn't we just remove the `= ...` here?

**Comment by Michael0x2a** [comment]
lol, good point -- fixed the title

File: stdlib/2/os/__init__.pyi (hunk 1 above)
Comment (Michael0x2a) [reply stdlib/2/os/__init__.pyi]: Done.

**Comment by Michael0x2a** [comment]
I reverted the changes to the attrs stubs for now -- it looks like the attrs stubs are currently impossible to type in a way that's compatible with the current and proposed overload semantics.

I submitted a pull request to slightly relax mypy's current overload semantics so that this change can go through. Once my other pull request is accepted, I'll unrevert my revert (unless you decide to accept this pull request as-is -- if so, I can just make another pull request for attrs).

**Comment by Michael0x2a** [comment]
@JelleZijlstra -- Ok, looks like third time was the charm. I think this is ready for a second look whenever you have time.

**Pull request merged**